### ISSUE TITLE
Add the ability to attach handling to an existing express app

### ISSUE THREAD

**Issue by EdwardSalter** [jormaechea/open-api-mocker#22]
At current this project is designed to run as a standalone application through the use of the embedded express server.
There are use cases where this project could be beneficial to be used within an existing express application or be run in a serverless architecture where a server is not required.

I suggest that a new server type be added that is designed to be used when accessing open-api-mocker via package only (i.e. it would throw an error if used by the CLI). This would be very similar to the [existing express server implementation](https://github.com/jormaechea/open-api-mocker/blob/master/lib/mocker/express/server.js) but instead of the `init` function starting up a server by listening on a port, a function is returned instead.
This function is then called in consuming code and is passed an existing express app.
Alternatively, one could pass the express app into the `mock` method of `OpenApiServer`.
It would look something like this:
```js
// Consuming code in third party application
const app = express();

  const server = new OpenApiMocker({
    port: undefined,
    server: "express-connector",
  });
server.validate()
  .then(() => server.mock())
  .then((connect) => {
      // The `init` method of the server (which is called in within the mock call above) returns a function that when called, adds the mocking functionality to the passed express app.
     connect(app);
  })
```

To avoid the consumer having to write their own code to handle watching the schema file for changes and attempting to reinitialise everything, I also propose a new "entry point" to be added as well that acts in a similar manner to the `open-api-mocker-cli.js` file (and shares some code with it).
It would be used like this:
```js
// Third party code
import connector from "open-api-mocker/express-connector";
import express from "express";
// or perhaps
// import express from "serverless-express/express";

const app = express(); // existing express app

const definitionFilePath = path.resolve(process.cwd(), "api-defs.yaml");
const { connect, schema } = connector({
  schemaPath: definitionFilePath,
  watch: process.env.NODE_ENV === "development",
});

connect(app)
  .then(() => {
    console.log("App connected successfully");
  })
  .catch((e) => {
  console.error("Failed to connect mock API", e);
});
```

**Comment by EdwardSalter** [comment]
I have attached a draft PR to give you a clearer idea if my explanations were not adequate.
There is still a bit of work to do around updating the branch, tidying up and moving shared code that is currently duplicated.
There are also broken tests that need fixing but I can sort those out relatively easily.

Any suggestions you have on this would be greatly appreciated.

**Comment by jormaechea** [comment]
Hey! Thanks for the issue. When you talked about this idea I was thinking more of some dependency injection that defaults to current implementation. Let me guide you through this:

First of all, the entry point [OpenApiMocker](https://github.com/jormaechea/open-api-mocker/blob/master/lib/open-api-mocker.js) is ready for some _server swapping_.
In my opinion, a small tweak should do. Imagine changing the constructor like this:
```js
const defaultServer = require('./mocker/express/server');

class OpenApiMocker {
	constructor({ port, server = defaultServer }) {
		this.server = new server();
		this.port = port || this.constructor.defaultPort;
	}
}
```

That way, by default the behaviour does not change. But you can pass your own `Server` implementation and do whatever you want in it.

Taking a quick look into the `Server` class, any other server must be implemented as a Class with an interface like this:

```js
class MyCustomServer {
  setServers(servers: OpenAPIServer[]): void;
  setPort(port: number): void;
  setPaths(paths: Paths[]): void;
  init(): void;
  shutdown(): void;
}
```

Other servers can even be distributed as separate npm packages to cover common use cases. And I see no reason to say that this won't work on the CLI.

What do you think?

**Comment by EdwardSalter** [comment]
I noticed the dependency injection and have already implemented a new server in [my draft PR](https://github.com/jormaechea/open-api-mocker/blob/d5e7dd79dcfa8fb0bb17d569cc01227ed08f774d/lib/mocker/express-connector/server.js). This implements the interface that you have outlined above with the difference being that `init()` returns a function in my Server implementation. I feel like the tweak you have suggested to the OpenApiMocker constructor is better option and I will look into making that change as it gives more control over how that server is initialised.

I like the idea of supporting different server implementations through npm packages as it promotes a plugin based architecture. Would this be your preferred option of implementing a new server? One thing I should point out however is that since this new server I have drafted is using express as well, just using an existing instance of creating a new one, I have modified the original express server to inherit this new class. All of the middleware connection is done in the new server and the existing one now simply creates the new express app before passing it to the superclass to connect and starting the server itself like it did before.

As for the additional entry point, I added the new express-connector entry point as a stop gap since the file watching functionality currently only exists on the open-api-mocker-cli file.
I'm thinking that perhaps we could move that functionality into the OpenApiMocker class itself so that it can be used by the CLI and through code requiring open-api-mocker as a dependency. Let me know your thoughts on that?

**Comment by jormaechea** [comment]
I wouldn't create any other server inside OpenAPI Mocker. Just expose the API to allow people to create their own.

Regarding the `--watch` feature, I agree: this should not be in the CLI layer. As you said, it could be moved to the `OpenApiMocker` class so that it can be used through code too.
Maybe we could add a `reload()` method to the Server interface that will called on each watched change (right now it calls `init` on startup and on each change, which I don't think is correct). It's not mandatory but it looks like a nicer/cleaner API.

`OpenApiMocker.mock()` logic will be something like this:
```js
async mock() {
	// Until init it does not change
	await this.server
		.setServers(this.api.servers)
		.setPort(this.port)
		.setPaths(this.api.paths)
		.init();

	if(hasToWatch) {
		watch().onChange(() => this.server.reload());
	}
}
```

What do you think?

**Comment by EdwardSalter** [comment]
I like the idea of a `reload()` method. I think that looks nicer in terms of an API.

In terms of the watch feature, I was thinking along the same kind of lines as the DI for the server; we could do the same thing with file watching and optionally inject a file loader/watcher implementation into the constructor or fall back to the default implementation. This would allow third party code to implement their own (I was considering previously the need to load a schema from a network path and poll for changes).

I am not quite sure how I can achieve this with the current API without making any breaking changes.
At current the `.setSchema` method is the sole means of getting the schema into the `OpenApiMocker` instance.
In terms of using DI for the file loading, it would look something like this I guess using a default file loading implementation:
```js
class DefaultFileLoader extends EventEmitter {
       	setOptions(options) {
		// Options could be any type, dependent on FileLoader implementation.
		// For DefaultFileLoader, options contains file path and watch flag.
	}

	load() {
		// Loads the schema and fires loaded event each time the file has loaded (i.e. if watching)
		const schema = {}; // Real code omitted for brevity
		this.emit('loaded', schema);
	}
}

class OpenApiMocker {
	constructor({ port, server, fileLoader }) {
		this.fileLoader = fileLoader || new DefaultFileLoader();
	}

        // Add this method since DefaultFileLoader will need options but don't want to add those to OpenApiMocker constructor if you are providing a means of injecting a loader. It also wouldn't be particularly user friendly to have to import the default file loader yourself and initialise a new instance passing the options that way. Still not 100% sure about this.
	setFileLoaderOptions(options) {
		this.fileLoader.setOptions(options);
		return this;
	}

	mock() {
		// Everything from before

		this.fileLoader.on('load', async schema => {
			this.setSchema(schema);
			await this.validate();

			return this.server.reload();
		});

		this.fileLoader.load();
	}
}


// consuming code
new OpenApiMocker({})
	.setFileLoaderOptions({ filePath: './schema.yaml', watch: true })
	.mock();
```

Do you think this would work?


With respect to the third party server implementations, you would see those packages as returning a Server instance only and then passing it to the OpenApiMocker constructor? Something like this:
```js
import OpenApiMocker from 'open-api-mocker';
import ExpressConnectorServer from 'open-api-mocker-express-connector-server';

const server = new ExpressConnectorServer(expressConnectorServerOptions);
const mocker = new OpenApiMocker({ server });
```

**Comment by jormaechea** [comment]
I'm not sure about the `setFileLoaderOptions()` part. Why don't you start with the `Server` DI, while I do a little bit of refactoring that I wanted to do for a while that should simplify much of the default vs custom schema loader.

I will open a separate issue to track this feature separately.

---

I totally agree with the custom Server implementation you just showed, let's get that done!

**Comment by jormaechea** [comment]
Resolved in #30
